velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp

```diff
@@ -121,6 +121,40 @@ class DateTimeFunctionsTest : public functions::test::FunctionBaseTest {
         unpackZoneKeyId(timestampWithTimezone)};
   }
 
+  std::optional<TimestampWithTimezone> atTimezone(
+      std::optional<int64_t> timestamp,
+      const std::optional<std::string>& timeZoneName,
+      const std::optional<std::string>& newTimeZoneName) {
+    facebook::velox::VectorPtr resultVector;
+    if (!timestamp.has_value() || !timeZoneName.has_value()) {
+      resultVector = evaluate(
+          "at_timezone(c0, c1)",
+          makeRowVector(
+              {BaseVector::createNullConstant(
+                   TIMESTAMP_WITH_TIME_ZONE(), 1, pool()),
+               makeNullableFlatVector<std::string>({newTimeZoneName})}));
+    } else {
+      resultVector = evaluate(
```

Comment:
```
evaluateOnce<int64_t>(
   "at_timezone(c0, c1)", 
   {timestamp, timeZoneName}, 
   {TIMESTAMP_WITH_TIME_ZONE(), VARCHAR()})
```

should work. Does it not work?

Reply:
Setting the return type to `int64_t` mitigates one issue, but that still doesn't work because `evaluateOnce()` cannot take in a `timestamp with time zone` type.

The equivalent approach requires using `evaluateWithTimestampWithTimezone()`, which calls `evaluate()` under the hood. But this function is not equipped to take in any additional arguments with a `timestamp with time zone`, which is why I made my custom function.

Though now that I think about it, I should probably just extend the existing `evaluateWithTimestampWithTimezone()` to perhaps have it take in additional optional arguments.

Reply:
There is this version of evaluateOnce, which on the surface is supposed to work, but .... Args is a single template parameter, not variadic like in the other version of evaluateOnce, hence, only works if there is a single argument or all arguments are of the same type, which is not our case.

```
  template <typename ReturnType, typename Args>
  std::optional<ReturnType> evaluateOnce(
      const std::string& expr,
      const std::vector<std::optional<Args>>& args,
      const std::vector<TypePtr>& types,
      const std::optional<SelectivityVector>& rows = std::nullopt,
      const TypePtr& resultType = nullptr) {
    std::vector<VectorPtr> flatVectors;
    for (vector_size_t i = 0; i < args.size(); ++i) {
      flatVectors.emplace_back(makeNullableFlatVector(
          std::vector<std::optional<Args>>{args[i]}, types[i]));
    }
    auto rowVectorPtr = makeRowVector(flatVectors);
    return evaluateOnce<ReturnType>(expr, rowVectorPtr, rows, resultType);
  }
```

I assume the following works:

```
evaluateOnce<int64_t>(
   "at_timezone(c0, c1)", 
   {ts1, ts2}, 
   {TIMESTAMP_WITH_TIME_ZONE(), TIMESTAMP_WITH_TIME_ZONE()})
```

Reply:
We can introduce a version of evaluateOnce that takes variadic number of arguments + a list of types:

```
  template <typename ReturnType, typename... Args>
  std::optional<ReturnType> evaluateOnce(
      const std::string& expr,
      const std::vector<TypePtr>& types,
      const std::optional<Args>&... args) {
    std::vector<VectorPtr> flatVectors;

    int i = 0;
    (
        [&] {
          flatVectors.emplace_back(makeNullableFlatVector(
              std::vector<std::optional<Args>>{args}, types[i]));
          ++i;
        }(),
        ...);

    return evaluateOnce<ReturnType>(expr, makeRowVector(flatVectors));
  }
```

Then, use it like

```
evaluateOnce<int64_t>(
        "at_timezone(c0, c1)", 
        {TIMESTAMP_WITH_TIME_ZONE(), VARCHAR()},
        timestamp,
        timeZoneName);
```

Reply:
@pedroerp @mbasmanova I rewrote the tests here, using the new `evaluateOnce()` framework. Certainly nice to avoid having to deal with the vectors directly, thank you Pedro! 

Please let me know if everything looks good here.